superdesk/publish_async/filters/base_exchange_filter.py

```diff
@@ -0,0 +1,546 @@
+        """
+
+        self.cache_global_filter_matches(request)
+        subscribers = await self.get_subscribers(request)
+        self.process_products(request, response, subscribers)
+        for subscriber in subscribers:
+            self.process_subscriber(request, response, subscriber)
+
+        if not response.subscribers:
+            logger.warning(f"No subscribers found for {request.item.get('id')}")
+            return
+
+    async def get_subscribers(self, request: PublishRequest) -> list[SubscribersResource]:
+        """
+        Retrieve matching subscribers for a given publish request.
+
+        This method filters cached subscribers based on the criteria defined in
+        the helper methods `subscriber_type_matches_request_type` and
+        `subscriber_target_matches_item_target`. It retrieves the subscribers from
+        a cached instance of `PublishCache` and applies the filtering logic to return
+        a list of relevant subscribers for the publish request.
+
+        :param request: The PublishRequest instance with details for filtering subscribers and products.
+        :return: A list of Subscribers to use for checking against this publish request.
+        """
+
+        cache = PublishCache.get()
+        return [
+            subscriber
+            for subscriber in cache.subscribers.values()
+            if (
+                self.subscriber_type_matches_request_type(request, subscriber)
+                and self.subscriber_target_matches_item_target(request, subscriber)
+            )
+        ]
+
+    def process_products(
+        self, request: PublishRequest, response: PublishRequestResponse, subscribers: list[SubscribersResource]
+    ) -> None:
+        """
+        Processes products and Content API products based on the provided request and the list of subscribers.
+
+        This method determines the set of product IDs and API product IDs from the given subscribers,
+        matches them to the request, and populates the response with the matched products,
+        matched API products, and their corresponding product codes.
+
+        :param request: The PublishRequest instance with details for filtering subscribers and products.
+        :param response: The response object to hold matched products, API products, and their product codes.
+        :param subscribers: A list of subscriber resources contributing to the product and API product IDs.
+        """
+
+        product_ids: set[ObjectId] = set(
+            chain(*[subscriber.products for subscriber in subscribers if subscriber.products])
+        )
+        api_product_ids: set[ObjectId] = (
+            set()
+            if not request.publish_to_content_api or not content_api.is_enabled()
+            else set(chain(*[subscriber.api_products for subscriber in subscribers if subscriber.api_products]))
+        )
+
+        response.matched_products = self.get_matching_products(request, product_ids)
```

Comment:
I think it would be useful to store matched products on the item itself, we could later display how many items were published per product, and it could be also used by formatters